transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollChannel.java

```diff
@@ -304,8 +311,47 @@ protected final int doWriteBytes(ByteBuf buf, int writeSpinCount) throws Excepti
         /**
          * Called once EPOLLRDHUP event is ready to be processed
          */
-        void epollRdHupReady() {
-            // NOOP
+        final void epollRdHupReady() {
+            if (isActive()) {
+                // If it is still active, we need to call epollInReady as otherwise we may miss to
+                // read pending data from the underlying file descriptor.
+                // See https://github.com/netty/netty/issues/3709
+                epollInReady();
+
+                // Clear the EPOLLRDHUP flag to prevent continuously getting woken up on this event.
+                clearEpollRdHup();
+            }
+            // epollInReady may call this, but we should ensure that it gets called.
+            shutdownInput();
+        }
+
+        /**
+         * Clear the {@link Native#EPOLLRDHUP} flag from EPOLL, and close on failure.
+         */
+        private void clearEpollRdHup() {
+            try {
+                clearFlag(Native.EPOLLRDHUP);
+            } catch (IOException e) {
+                pipeline().fireExceptionCaught(e);
+                close(voidPromise());
+            }
+        }
+
+        /**
+         * Shutdown the input side of the channel.
+         */
+        void shutdownInput() {
+            if (!inputShutdown) { // Best effort check on volatile variable to prevent multiple shutdowns
```

Comment:
@normanmaurer - I could change this, however I don't think there is any requirement to prevent multiple shutdowns.

For example see [AbstractNioByteChannel](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java#L62).

If we want to be sure (use CAS) I would suggest that I remove this check here and submit a followup PR. WDYT?